### DISCUSSION TITLE
How can I tell what U(S)ART is used with setTx() and setRx() when more than one is possible?

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#2319] [Q&A] [answered]
Using for example STM32L4R5, setting serial Rx and Tx to PC11 and PC10
When I check in the datasheet, on those pins I have USART3 or UART4
PA3 and PA2 can be LPUART1 or USART2

How do I know what peripheral is beeing used?

Also, is there a difference in using setTx() and setRx() and defining a new HardwareSerial using HardwareSerial newSerial(Rx, Tx)?

**Comment by fpistm** [comment, marked as answer]
Hi @MasteringTheMess 
You can find the information in the PeripheralPins.c, all possible value are referenced here:
https://github.com/stm32duino/Arduino_Core_STM32/blob/8f337e8aa4ce0eee00c75b15102b81e09e1b2b6e/variants/STM32L4xx/L4R5Z(G-I)T_L4R7ZIT_L4S5ZIT_L4S7ZIT/PeripheralPins.c#L208-L209

To choose the right instance you can use the _ALTx notation.
Example to use `USART2`
```C++
mySerial.setTx(PA2_ALT1);
mySerial.setRx(PA3_ALT1);
```
while using:
```C++
mySerial.setTx(PA2);
mySerial.setRx(PA3);
```
will use `LPUART1`.

Note that the PeripheralPins.c reference the `PinName` (`PY_n`). Arduino API use the pin number (`PYn`).

**Comment by PerennialNovice** [comment]
Thanks!

...so, using the Arduino pin numbers will choose the first instance that is found in the PeripheralPins.c file?

**Reply by fpistm** [reply]
No. It takes the one specified by the pin number you used. If you use PA2 it uses LPUART1, if you use PA2_ALT1 it uses LPUART. I thought my answer was clear...

**Reply by PerennialNovice** [reply]
ok, so PA2 and PA_2 are the same and the Arduino API does not know about the ALT pins...

**Reply by fpistm** [reply]
PA2 is the pin number while PA_2 is a PinName ( combination of GPIO port and pin). PA2 is simply a define to a number (index) to be able to retrieve the correct PinName PA_2. PA2_ALT1 allows to retrieve PA_2_ALT1.

**Reply by PerennialNovice** [reply]
thanks a lot for explaining so patiently :)
